chain/default_syncer.go

```diff
@@ -88,82 +94,63 @@ func NewDefaultSyncer(online, offline *hamt.CborIpldStore, c consensus.Protocol,
 // TODO the timeout factor blkWaitTime and maybe the whole timeout mechanism
 // could use some actual thought, this was just a simple first pass.
 func (syncer *DefaultSyncer) getBlksMaybeFromNet(ctx context.Context, blkCids []cid.Cid) ([]*types.Block, error) {
-	var blks []*types.Block
 	ctx, cancel := context.WithTimeout(ctx, blkWaitTime)
 	defer cancel()
-	for _, blkCid := range blkCids {
-		// try the chain store
-		blk, err := syncer.chainStore.GetBlock(ctx, blkCid)
-		if err == nil {
-			blks = append(blks, blk)
-			continue
-		}
-		// try the node's local offline storage
-		err = syncer.cstOffline.Get(ctx, blkCid, &blk)
-		if err == nil {
-			blks = append(blks, blk)
-			continue
-		}
-		// try the network
-		if err = syncer.cstOnline.Get(ctx, blkCid, &blk); err != nil {
-			return nil, err
-		}
-		blks = append(blks, blk)
-	}
-	return blks, nil
+
+	return syncer.fetcher.GetBlocks(ctx, blkCids)
 }
 
-// collectChain resolves the cids of the head tipset and its ancestors to blocks
-// until it resolves blocks contained in the Store. collectChain may resolve cids
-// from the Store, the node's local offline cborstore, or the syncer's online
-// cbor store that is networked under the hood. collectChain errors if any
-// set of cids in the chain resolves to blocks that do not form a tipset, if
-// the chain is too long, or if any tipset has already been recorded as the
-// head of an invalid chain.
-//
-// collectChain is the entrypoint to the code that interacts with the network.
-// It does NOT add tipsets to the store.
-func (syncer *DefaultSyncer) collectChain(ctx context.Context, blkCids []cid.Cid) ([]types.TipSet, types.TipSet, error) {
+// collectChain resolves the cids of the head tipset and its ancestors to
+// blocks until it resolves a tipset with a parent contained in the Store. It
+// returns the chain of new incompletely validated tipsets and the id of the
+// parent tipset already synced into the store.  collectChain resolves cids
+// from the syncer's fetcher.  In production the fetcher wraps a bitswap
+// session.  collectChain errors if any set of cids in the chain resolves to
+// blocks that do not form a tipset, or if any tipset has already been recorded
+// as the head of an invalid chain.  collectChain is the entrypoint to the code
+// that interacts with the network. It does NOT add tipsets to the store.
+func (syncer *DefaultSyncer) collectChain(ctx context.Context, blkCids []cid.Cid) ([]types.TipSet, string, error) {
 	var chain []types.TipSet
 	defer logSyncer.Info("chain synced")
 	for {
 		var blks []*types.Block
 		// check the cache for bad tipsets before doing anything
 		tsKey := types.NewSortedCidSet(blkCids...).String()
 
+		// Finish traversal if the tipset made is tracked in the store.
+		if syncer.chainStore.HasTipSetAndState(ctx, tsKey) {
```

Comment:
During testing I discovered that the fetcher needed to be able to fetch blocks that were already in the chainStore for the syncer to work right.  This was indirect and unexpected so I moved this check up and changed the function signature to return a tipset id that the caller can use to get the parent tipset from the store.

Reply:
Thanks, that's a great idea.